### ISSUE TITLE
microdom.toxml mishandles xml namespaces

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2184]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2184|
|Type|defect|
|Created|2006-10-24 01:09:53Z|

Noted in http://divmod.org/trac/ticket/1621

```
>>> microdom.parseString('''
... <html xmlns="http://www.w3.org/TR/REC-html40">
... <body xmlns="">
... </body>
... </html>
... ''').toxml()
Traceback (most recent call last):
  File "<stdin>", line 1, in ?
  File "/Users/jml/Projects/Twisted/trunk/twisted/web/microdom.py", line 113, in toxml
  File "/Users/jml/Projects/Twisted/trunk/twisted/web/microdom.py", line 212, in writexml
  File "/Users/jml/Projects/Twisted/trunk/twisted/web/microdom.py", line 488, in writexml
  File "/Users/jml/Projects/Twisted/trunk/twisted/web/microdom.py", line 440, in writexml
KeyError: 'http://www.w3.org/TR/REC-html40'
```

<details><summary>Searchable metadata</summary>

```
trac-id__2184 2184
type__defect defect
reporter__jml jml
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1161652193000000 1161652193000000
changetime__1162069490000000 1162069490000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml removed owner|
|-|-|

Specifically, microdom doesn't allow the default namespace to become unspecified, as in http://www.w3.org/TR/REC-xml-names/#scoping-defaulting

Up for review in xmlns-2184

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @dreid|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_3"></a>|@dreid set owner to @jml|
|-|-|

Ok, this was an easy one.  The tests pass and there is nothing offensive about the code changes.  Some other testcases in test_xml seem to violate the naming standard but that's another ticket.

Ok to merge.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_4"></a>|@jml set status to `closed`|
|-|-|

(In [18513]) Allow the default namespace to 'undefaulted' in microdom
 * Fixes [#2184](https://github.com/twisted/twisted/issues/2184)
 * Author: jml
 * Reviewer: dreid

According the W3C, the specifying xmlns="" in an element means that 
within the element, there is no default namespace. See 
http://www.w3.org/TR/REC-xml-names/#scoping-defaulting

Previously, microdom would raise an exception when it encountered a thing
like this.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml set status to `reopened`|
|-|-|

OK, I've fixed the lore failures.  Ready for review again.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @dreid<br>@jml set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_7"></a>|@jerub set owner to @jml|
|-|-|

you forgot to mark it for review.

Please merge. :)

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_8"></a>|@jerub commented|
|-|-|

or maybe I'm insane!

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_9"></a>|@jml set status to `closed`|
|-|-|

(In [18550]) Allow the default namespace to be 'undefaulted' in microdom.
 * Fixes [#2184](https://github.com/twisted/twisted/issues/2184)
 * Author: jml
 * Reviewers: dreid, Jerub

According the W3C, the specifying xmlns="" in an element means that
within the element, there is no default namespace. See
http://www.w3.org/TR/REC-xml-names/#scoping-defaulting

Previously, microdom would raise an exception when it encountered a thing
like this. Now, microdom knows the differences between "no namespace 
specified" and "default namespace removed".

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
